pom.xml

```diff
@@ -131,15 +132,15 @@
                     <!-- Remove the path from the root directory until current sourceDirectory to obtain a relative path -->
                     <script language="beanshell"><![CDATA[
                       String path = project.getProperty("file_path");
-                      String source = project.getProperty("docgen.asciidoc.source");
-                      path = path.replace(source, "");                      
+                      String source = project.getProperty("basedir");
```

Comment:
basedir is not possible. I had to change it intentionally to `docgen.asciidoc.source` that defaults to `basedir`.
Is there a particular reason why you reverted this change?

Reply:
I reverted it because I was getting an empty value when using getProperty("docgen.asciidoc.source") instead of "basedir"

Reply:
Today I'm doing some tests with `docgen.asciidoc.source` and in fact, I'm getting the correct value. I guess that I was getting that "empty value" error from a different source.

Reply:
![image](https://user-images.githubusercontent.com/43032659/72055417-16c54b80-32cb-11ea-9bc3-5fad157d6407.png)


When you specify `docgen.asciidoc.source`, then I get errors in this method because I use `basedir` in order to remove it's value from `file_path` in order to obtain a relative path.

Basically if:
* basedir is C://a/b
* docgen.asciidoc.source is   C://a/b/components
* file_path is C://a/b/c/d

Then, using ´basedir´ as source will allow to obtain `/c/d` (replace method), but `C://a/b/components` is not a substring of `file_path`, so I won't obtain `/c/d` (replace method would return `C://a/b/c/d`)

Reply:
Thanks for your concrete feedback and investigation.
I still do not get why in your case `docgen.asciidoc.source` is not the same as `basedir`.
In your POM do you have `docgen.asciidoc.source` set to something other than the following?
```
<docgen.asciidoc.source>${basedir}</docgen.asciidoc.source>
```

> docgen.asciidoc.source is C://a/b/components

That would IMHO mean that docgen is only processing asciidoc from this directory.
How can then be that:

> file_path is C://a/b/c/d

Reply:
To give a motivation/explanation:
So far `${basedir}` was hardcoded to be the directory to process the asciidoc from.
When you have modular builds with maven sub-modules inside the asciidoc generation you get into trouble as then the asciidoc processing will also recursively travers the sub-modules folders. In the worst case this can include target folders what increases the number of files exponentially with every build if no clean is performed.
Therefore I added `devon.asciidoc.source` as a new feature to allow the user to override the default of `${basedir}` to someting else (e.g. `${basedir}/asciidoc` or something).
IMHO this requires to use that directory from `devon.asciidoc.source` also in the ant-processing for the relative paths. Maybe we can investigate together what exactly went wrong here and how to fix this properly. I am one of those people who want to fully understand what is happening before fixing something to prevent that some best guess fix turns out to be totally wrong later.

Reply:
> That would IMHO mean that docgen is only processing asciidoc from this directory.
> How can then be that:
> 
> > file_path is C://a/b/c/d

You are right, It should be "file_path is C://a/b/components/d"

Here you have an example extracted from my specific situation:

```

------------- basedir --------------
C:/Proyectos/new-docgen-tst/devonfw.github.io/website
------------- docgen.asciidoc.source --------------
C:/Proyectos/new-docgen-tst/devonfw.github.io/website/components/
------------- file_path --------------
C:/Proyectos/new-docgen-tst/devonfw.github.io/website/components/footer

```

`docgen.asciidoc.source` : ${basedir}/components/
`docgen.asciidoc.target`: ${basedir}/target/asciidoc/
`docgen.generated.docs`: ${basedir}/target/generated-docs/

* If I use `docgen.asciidoc.source` then 
  * the result of the pre-processing will go to `C:\Proyectos\new-docgen-tst\devonfw.github.io\website\target\asciidoc\footer`
Notice that there is no `/components/` between `C://.../target/asciidoc` and `/footer` due to string.replace(...) 
  * The obtained htmp will be output to `C:\Proyectos\new-docgen-tst\devonfw.github.io\website\target\generated-docs\footer` (no `/components/`)

* If I use `basedir` then
  * the result of the pre-processing will go to `C:\Proyectos\new-docgen-tst\devonfw.github.io\website\target\asciidoc\components\footer` 
  * The obtained htmp will be output to `C:\Proyectos\new-docgen-tst\devonfw.github.io\website\target\generated-docs\components\footer`

So, my question is: Should then source path be preserved in the destination? If I'm in ${basedir}/components should the output be in  ${basedir}/target/asciidoc/components/ and ${basedir}/target/generated-docs/components/ or should it be specified like:

`docgen.asciidoc.source` : ${basedir}/components/
`docgen.asciidoc.target`: ${basedir}/target/asciidoc/components
`docgen.generated.docs`: ${basedir}/target/generated-docs/components

...?

|                              | basedir                                                   | docgen.asciidoc.source                                    |
|------------------------------|-----------------------------------------------------------|-----------------------------------------------------------|
| Current value                | C:/.../website/                                           | C:/.../website/components/                                |
| file_path                    | C:/.../website/components/footer                          | C:/.../website/components/footer                          |
| file_path.replace(value, "") | components/footer                                         | footer                                                    |
| preprocessing output path    | C:\...\website\target\asciidoc\ + components\footer       | C:\...\website\target\asciidoc\ + footer                  |
| pdf/html output path         | C:\...\website\target\generated-docs\ + components\footer | C:\...\website\target\generated-docs\ + components\footer |

Reply:
Hi @jambulud - thanks for taking the time to write all the details.
So we not talking about a bug but more about the design/behaviour of a new feature that I introduced with my PR/change.
So if you change `docgen.asciidoc.source` from `${basedir}` to something else (e.g. a sub-folder) I assume that you do not care about the initial structure containing the sub-folder. Therefore the behaviour of 1. is exactly as designed. I fully agree that the behaviour of 2. with a duplicated "components/components" is kind of a bug. Is that required because for devonfw website we are relying on specific relative paths in includes or hyperlinks? Why do we need to change `docgen.asciidoc.source` then in that case?

Please note that I am blocked with further progress in various repos until we have the new docgen released and this discussion is further blocking to make that release going to happen. If that new feature is really a problem, I would propose to take it out entirely and create a new issue for it with all this discussion so we move that feature to an upcoming release. The other alternative would be to consider 2. as a bug and file an issue for it so I can try fix that after this merge is complete but unless this PR is reworked, I am kind of stuck. We have to come to some agility to be able to make progress.

Reply:
1 and 2 are different configurations and its effect with using basedir or docgen.asciidoc.source in the beanshell script.

If you want to use `docgen.asciidoc.source` in the script instead of `basedir`, then the given configuration in 2 needs to be used in order to work correctly. If the behaviour stated in 2 for `docgen.asciidoc.source` is fine for you, then you can merge this PR. I have done tests with that particular configuration and I get the website without errors.

Reply:
I added docgen.asciidoc.source again and added `.replace("\\", "/")` in order to get rid of the backslashes because in pom files paths are usually written using "/" so in windows I get things like C://a\b\c/path/in/pom/